qiskit_machine_learning/algorithms/effective_dimension.py

```diff
@@ -0,0 +1,337 @@
+            model_outputs = np.expand_dims(model_outputs, axis=2)
+
+        # get grad-vectors (gradient_k/model_output_k)
+        # multiply by sqrt(model_output) so that the outer product cross term is correct
+        # after Einstein summation
+        gradvectors = np.sqrt(model_outputs) * gradients / model_outputs
+
+        # compute sum of matrices obtained from outer product of grad-vectors
+        fishers = np.einsum("ijk,lji->ikl", gradvectors, gradvectors.T)
+
+        return fishers
+
+    def get_fhat(self, fishers: np.ndarray) -> Tuple[np.ndarray, float]:
+        """
+        This method computes the normalized Fisher Information Matrix (f_hat)
+        and extracts its trace.
+        Args:
+            fishers: The Fisher Information Matrix to be normalized.
+        Returns:
+             f_hat: The normalized FIM, a numpy array of size (num_inputs, d, d)
+             fisher_trace: The trace of the Fisher Information Matrix
+                            (before normalizing).
+        """
+
+        # compute the trace with all fishers
+        fisher_trace = np.trace(np.average(fishers, axis=0))
+
+        # average the fishers over the num_inputs to get the empirical fishers
+        fisher_avg = np.average(
+            np.reshape(
+                fishers,
+                (
+                    self.num_params,
+                    self.num_inputs,
+                    self._model.num_weights,
+                    self._model.num_weights,
+                ),
+            ),
+            axis=1,
+        )
+
+        # calculate f_hats for all the empirical fishers
+        f_hat = self._model.num_weights * fisher_avg / fisher_trace
+        return f_hat, fisher_trace
+
+    def _get_eff_dim(
+        self, f_hat: Union[List[float], np.ndarray], n: Union[List[int], np.ndarray, int]
+    ) -> Union[np.ndarray, int]:
+
+        if not isinstance(n, int) and len(n) > 1:
+            # expand dims for broadcasting
+            f_hat = np.expand_dims(f_hat, axis=0)
+            n_expanded = np.expand_dims(np.asarray(n), axis=(1, 2, 3))
+            logsum_axis = 1
+        else:
+            n_expanded = n # type: ignore
+            logsum_axis = None
+
+        # calculate effective dimension for each data sample size "n" out
+        # of normalized fishers
+        f_mod = f_hat * n_expanded / (2 * np.pi * np.log(n_expanded))
+        one_plus_fmod = np.eye(self._model.num_weights) + f_mod
+        # take log. of the determinant because of overflow
+        dets = np.linalg.slogdet(one_plus_fmod)[1]
+        # divide by 2 because of square root
+        dets_div = dets / 2
+        effective_dims = (
+            2
+            * (logsumexp(dets_div, axis=logsum_axis) - np.log(self.num_params))
+            / np.log(n / (2 * np.pi * np.log(n)))
+        )
+
+        return np.squeeze(effective_dims)
+
+    def get_eff_dim(self, n: Union[List[int], np.ndarray, int]) -> Union[np.ndarray, int]:
```

Comment:
I'll let other people comment on names. For me though, so things are more self documenting why not `get_effective_dimension`. I think get_effective_dim might be ok too as dim is regularly used for dimension. get_fhat seems obtuse, And why is it get_fishers plural - that seems to be get fisher matrix so seems singular to me - the comment says Fisher Information Matrix so it seemed singular in respect of fisher.